[Security] Make login redirection logic available to programmatic login

hellomedia opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3 
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        |  [17556](https://github.com/symfony/symfony-docs/pull/17556)


The `Security::login()` method [introduced in 6.2](https://github.com/symfony/symfony/pull/41274/files) returns `void`, whereas `UserAuthenticator::authenticateUser()` called under the hood returns a response which can be used to redirect the user.

This PR makes `login()` return the response from `authenticateUser()`, allowing the regular login redirection logic to be used if needed.


**Usage**

````php

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Bundle\SecurityBundle\Security;
use Symfony\Component\HttpFoundation\Response;

class RegistrationController extends AbstractController
{    
    public function verifyUserEmail(Security $security): Response
    {

        ...

        $redirectResponse = $security->login($user);
    
        return $redirectResponse;
    }
}
````

hellomedia committed:
[Security] Make login redirection logic available to programmatic login

chalasr commented on src/Symfony/Bundle/SecurityBundle/Security.php:
```diff
@@ -51,8 +51,10 @@ public function getFirewallConfig(Request $request): ?FirewallConfig
      * @param string|null   $authenticatorName The authenticator name (e.g. "form_login") or service id (e.g. SomeApiKeyAuthenticator::class) - required only if multiple authenticators are configured
      * @param string|null   $firewallName      The firewall name - required only if multiple firewalls are configured
+     *
+     * @return Response|null The UserAuthenticator::authenticateUser Response if any
```
```suggestion
     * @return Response|null The authenticator success response if any
```

nicolas-grekas commented:
Can you add a small test case also?

hellomedia commented:
@chalasr @nicolas-grekas Thank you both for the feedback.
I changed the comment as suggested and added a test.

Unfortunately, I apparently messed up the PR base branch when trying to rebase it.

I folllowed these [rebasing instructions](https://symfony.com/doc/current/contributing/code/pull_requests.html#rebase-your-pull-request) :
```
git checkout 6.3
git fetch upstream
git merge upstream/6.3
git checkout BRANCH_NAME
git rebase 6.3
git push --force origin BRANCH_NAME
```
but it now looks like my PR includes all latest commits from the 6.3 branch.

I have tried a few things but no success :( Can you point me to an easy fix or should I submit a new PR ?

hellomedia commented:
Rebasing issue seems kind of fixed now.

Sorry for the noise and the extra review for requests (automated).

chalasr reviewed (commented):
Can you add an entry in SecurityBundle/CHANGELOG.md?

hellomedia commented:
> Can you add an entry in SecurityBundle/CHANGELOG.md?

Sure.
What is best ?

> Make login redirection logic available to programmatic login

or

> Make programmatic login return authenticator response

chalasr commented:
I suggest 
> Make `Security::login()` return the authenticator response `

hellomedia commented:
Perfect, thanks. Done.

nicolas-grekas commented:
Thank you @hellomedia.
